### ISSUE TITLE
Render problem

### ISSUE THREAD

**Issue by SCratORS** [SCratORS/Everchanging#38]
If you set the screen resolution to 240*320, you can notice artifacts of the display of objects. It is clearly visible on the "Dandelions". I understand what this is about, but I can't find a solution to how it can be improved.
<img width="388" alt="render_problem" src="https://github.com/SCratORS/Everchanging/assets/24729719/a1c0234f-b89b-4102-a06d-7b0eab2eb0c4">

**Comment by Doctoror** [comment]
I believe this is happening because we convert vector graphics into a Bitmap and then scale the Bitmap.
This should be fixable with multisampling:
http://www.saschahlusiak.de/2012/10/opengl-antialiasing-in-android-with-transparent-textures/

I tried enabling multisampling but it does not work on emulators, and I don't have 240x320 physical device to verify.

**Comment by Doctoror** [comment]
I believe this is happening because we convert vector graphics into a Bitmap and then scale the Bitmap.
This should be fixable with multisampling:
http://www.saschahlusiak.de/2012/10/opengl-antialiasing-in-android-with-transparent-textures/

I tried enabling multisampling but it is limited on emulators, and I don't have 240x320 physical device to verify.
Emulators only supports multisampling of 2 and it did not look like it changed anything.

I found out that emulaors might not even have multisampling working at all, even when you open Emulator menus
https://www.reddit.com/r/AndroidStudio/comments/qh3jf6/avd_resolution_antialiasing_problem/?rdt=64969

**Comment by SCratORS** [comment]
It is possible that it is a matter of multisampling. But it doesn't seem like it. You can use a larger resolution. for example, 540*960 effect is also observed. I'm not sure exactly, but probably BlueStack (as an emulator created for android games on PC) supports multisampling, but the effect persists.

**Comment by Doctoror** [comment]
But have you enabled multi sampling in the egl config chooser before retesting on blue stack?

**Comment by SCratORS** [comment]
I can't find any such settings there. there is only software, hardware, opengl or vulkan rendering

**Comment by Doctoror** [comment]
This has to be enabled in app. I will send you a branch to test later, in a few hours, when I have time.

**Comment by SCratORS** [comment]
ok

**Comment by Doctoror** [comment]
You can check with multisampling on this branch
https://github.com/Doctoror/Everchanging/tree/multisampling

I copied EGLConfigChoosers from my other repo hence the licenses and package names.
The logic behing the choosers is that a config chooser can fail if unsupported config is attempted to be chosen.
If multisampling with value 4 is not supported 2 will be attempted.
Nvidia multisampling will be attempted if default multisampling fails.
The config with null will be passed if all other configs failed.
Null config means that the platform requests no requirements.

**Comment by SCratORS** [comment]
I got this in the logcat:
> GL com.scrat.everchanging ViewportNeedFlush 0
> EGLConfig chosen: [12324, 8, 12323, 8, 12322, 8, 12325, 0, 12338, 1, 12337, 4, 12352, 4, 12344]

It did not have any visible effect

**Comment by Doctoror** [comment]
I thought the texture is generated with intrinsic width, and then scaled down by matrix transform 🤔

The texture itself isn't in vector graphics then, I suppose.

**Comment by SCratORS** [comment]
A texture is being generated from an XML Drawable - Canvas Picture. Using the drawable.draw(canvas) method [here](https://github.com/SCratORS/Everchanging/blob/247efe1bcc2d48dc87797ae0a9f8d7462b4ca7f5/app/src/main/java/com/scrat/everchanging/TextureManager.java#L89)

**Comment by Doctoror** [comment]
Tes, I saw that. What I mean is that I don't see any resizing occur there. That's why I don't get it why it looks scaled down.

BTW I see some problems with that code that unnecessarily loads bitmap twice and also uses id resolution based on names. It doesn't affect the scaling issue, of course, but that's something that can be improved. I will open a PR for that later.

**Comment by SCratORS** [comment]
The size problem is that the xml vector accepts dp (dip pixel) as units of length. because of this, the lower the DPI of the display, the smaller the texture itself at the output

**Comment by Doctoror** [comment]
> the lower the DPI of the display, the smaller the texture itself at the output.

I think you meant the smaller the dpi the bigger the texture. And the bigger it is the more OpenGL must scale it down.

> The size problem is that the xml vector accepts dp

Not necessarily. You can do this:

```xml
<vector xmlns:android="http://schemas.android.com/apk/res/android"
    android:width="180.9px"
    android:height="208.05px"
    ...
```

But I don't think that this is correct. You want smaller output on smaller density devices and bitter output on bigger density devices. That's why dp is useful. But I see that the loaded images are too big. For example, this is the original size of the dandelion

![Test](https://github.com/SCratORS/Everchanging/assets/6018890/f88fb613-978e-4fac-9127-7a032c69f60f)

**Comment by Doctoror** [comment]
I think multisampling doesn't work because we don't use framebuffers.
I don't know how to use them. I tried something here but couldn't make it work.

Commit:
https://github.com/SCratORS/Everchanging/commit/9ab3fd9d1400f19545c82ab9fe271aa455c7e962

Branch:
https://github.com/Doctoror/Everchanging/tree/framebuffers

**Comment by SCratORS** [comment]
>But I see that the loaded images are too big anyway.

Yes, the original had such dimensions. I didn't invent anything here myself.

**Comment by SCratORS** [comment]
If in the texture manager, where the texture of the vector object is generated, add this:
         
```
drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
drawable.draw(canvas);
drawable.setBounds(8, 0, canvas.getWidth()+8, canvas.getHeight());
drawable.draw(canvas);
drawable.setBounds(16, 0, canvas.getWidth()+16, canvas.getHeight());
drawable.draw(canvas);
drawable.setBounds(16, 8, canvas.getWidth()+16, canvas.getHeight()+8);
drawable.draw(canvas);
drawable.setBounds(16, 16, canvas.getWidth()+16, canvas.getHeight()+16);
drawable.draw(canvas);
drawable.setBounds(8, 16, canvas.getWidth()+8, canvas.getHeight()+16);
drawable.draw(canvas);
drawable.setBounds(0, 16, canvas.getWidth(), canvas.getHeight()+16);
drawable.draw(canvas);
drawable.setBounds(0, 8, canvas.getWidth(), canvas.getHeight()+8);
drawable.draw(canvas);
```

Of course, this helps to partially solve the problem, but this is a very bad option in my opinion.

**Comment by SCratORS** [comment]
![image](https://github.com/SCratORS/Everchanging/assets/24729719/ae4b24a5-6183-4de7-9abd-ac34b3e28809)

Do you also have such a strip on the right in the preview?

**Comment by Doctoror** [comment]
Yes, this doesn't seem like a good option. I think it's okay to have a bigger texture scaled down, we just need to make multi sampling work. It works with textures only with OpenGL ES 3.2. I'm not sure how to correctly request 3.2 and how to make frame buffers work.

In my config chooser I enable Open GL ES 2.0 which is wrong, I don't know how to request 3.2. Maybe just remove that line?

`EGL10.EGL_RENDERABLE_TYPE, EGL_OPENGL_ES2_BIT`

Also, the framebuffers apparently don't work correctly.
If you have time, try to figure this out.

Once we have texture multi sampling in framebuffers I believe it should look fine.

**Comment by SCratORS** [comment]
I figured out where to go. I will study the Framebuffer issue. It seems to be true.

**Comment by SCratORS** [comment]
No matter how much I try, nothing normal happens. Either nothing at all, Or the same thing, Or the same thing but with glitches.

**Comment by Doctoror** [comment]
I couldn't figure that as well. But I don't have much time anymore.

**Comment by SCratORS** [comment]
Apparently we should leave it as it is for now
